app/controllers/job_orders_controller.rb

```diff
@@ -0,0 +1,396 @@
+class JobOrdersController < ApplicationController
+  before_action :current_user
+  before_action :signed_in, except: [:user_magic_approval, :user_magic_approval_confirmation]
+  before_action :grant_access, only: %w[admin settings start_processing processed paid picked_up quote_modal timeline_modal decline_modal completed_email_modal quote resend_quote_email]
+  before_action :set_job_order, only: %w[steps destroy user_approval start_processing processed paid picked_up quote_modal timeline_modal decline_modal completed_email_modal quote invoice resend_quote_email]
+  before_action :wizard, only: %w[steps]
+  before_action :allow_edit, only: %w[steps]
+
+  def index
+    @job_orders = []
+    @archived_job_orders = []
+    @drafts = []
+    @user.job_orders.not_deleted.each do |jo|
+      if jo.job_order_statuses.last.job_status == JobStatus::DRAFT
+        @drafts << jo
+      elsif jo.job_order_statuses.last.job_status == JobStatus::DECLINED || jo.job_order_statuses.last.job_status == JobStatus::PICKED_UP
+        @archived_job_orders << jo
+      else
+        @job_orders << jo
+      end
+    end
+  end
+
+  def new
+    @job_order = JobOrder.new
+  end
+
+  def steps
+    error = false
+
+    if @job_order.present? && @job_order.user == @user
+
+      if params[:keep_files].present?
+        (@job_order.user_files.pluck(:id) - params[:keep_files].map(&:to_i)).each do |file_id|
+          @job_order.user_files.find(file_id).purge
+        end
+      elsif params[:should_delete_user_files] && !params[:keep_files].present?
+        (@job_order.user_files.pluck(:id)).each do |file_id|
+          @job_order.user_files.find(file_id).purge
+        end
+      end
+
+      if params[:job_order].present?
+
+        if params[:job_service_name].present?
+          @job_order.job_services << JobService.create!(name: params[:job_service_name], job_service_group_id: params[:job_order][:job_service_group_id], user_created: true, job_order_id: @job_order.id)
+        end
+
+        if params[:job_order][:job_service_ids] == "custom"
+          params[:job_order][:job_service_ids] = @job_order.job_services.last.id
+        end
+        unless @job_order.update(job_order_params)
+          error = true
+        end
+      end
+
+      if params[:change_options].present? && params[:change_options] == "true"
+        unless @job_order.add_options(params)
+          error = true
+        end
+      end
+    else
+      error = true
+    end
+
+    if error
+      redirect_to job_order_steps_path(@job_order, step: (params[:step].to_i - 1))
+      flash[:alert] = "An error occurred while saving the job order step. make sure that you uploaded the right file types. Please try again."
```

Comment:
is filetype the only reason this block can fail? If we not we can convert error from bool to string and in each block set what the error would be and append it to the flash at the end if its not empty

Reply:
No it isn't, I'll try to improve this a bit